src/Analyzer.Utilities/DiagnosticHelpers.cs

```diff
@@ -98,5 +99,117 @@ public static string GetMemberName(ISymbol symbol)
             // For other language constructs
             return symbol.Name;
         }
+               
+        public static bool IsInvisibleOutsideAssemblyAtSymbolLevel(ISymbol symbol)
+        {
+            return SymbolIsPrivateOrInternal(symbol)
+                || SymbolIsProtectedInSealed(symbol);
+        }
+
+        public static bool SymbolIsPrivateOrInternal(ISymbol symbol)
+        {
+            var access = symbol.DeclaredAccessibility;
+            return access == Accessibility.Private
+                || access == Accessibility.Internal
+                || access == Accessibility.ProtectedAndInternal
+                || access == Accessibility.NotApplicable;
+        }
+
+        public static bool SymbolIsProtectedInSealed(ISymbol symbol)
+        {
+            var containgType = symbol.ContainingType;
+            if (containgType != null && containgType.IsSealed)
+            {
+                var access = symbol.DeclaredAccessibility;
+                return access == Accessibility.Protected
+                    || access == Accessibility.ProtectedOrInternal;
+            }
+
+            return false;
+        }
+
+        public static string GetMeaningfulParentName(SyntaxNode cur, SemanticModel model)
+        {
+            while (cur.Parent != null)
+            {
+                var pNode = cur.Parent;
+                ISymbol sym = pNode.GetDeclaredOrReferencedSymbol(model);
+
+                if (sym != null &&
+                    !string.IsNullOrEmpty(sym.Name)
+                    && (
+                        sym.Kind == SymbolKind.Method ||
+                        sym.Kind == SymbolKind.NamedType
+                       )
+                )
+                {
+                    return sym.Name;
+                }
+
+                cur = pNode;
+            }
+
+            return String.Empty;
+        }
+
+/// <summary>
+/// Determine wether a type (given by name) is actually declared in the expected assembly (also given by name)
+/// </summary>               
+/// <remarks>
+/// This can be used to decide wether we are referencing the expected framework for a given type. 
+/// For example, System.String exists in mscorlib for .NET Framework and System.Runtime for other framework (e.g. .NET Core). 
+/// </remarks>
+public static bool? IsTypeDeclaredInExpectedAssembly(Compilation compilation, string typeName, string assemblyName)
+        {
+            if (compilation == null)
+            {
+                return null;
+            }
+            var typeSymbol = compilation.GetTypeByMetadataName(typeName);
+            return typeSymbol?.ContainingAssembly.Identity.Name.Equals(assemblyName, StringComparison.Ordinal);
```

Comment:
What if "GetTypeByMetadataName"  returns null? Caller will not be able to tell the difference.